### PR TITLE
Mask interpolate

### PR DESCRIPTION
**Pull request by class4kayaker**
Work on a fix to #2811.

Due to a primary expected use case being to interpolate a function that only defines specified components into a system, requiring that the provided function return values for all of the component is not desirable. Currently, the implementation assumes that the components returned by the function match the order of the components within the system, skipping any of the components that should not be interpolated. This implementation does not currently permit interpolation only interpolating components of an FE system that also includes non-interpolating components.

### CODE REVIEW

**Comment by tamiko** [comment]
@class4kayaker This is work in progress?

**Comment by bangerth** [comment]
@class4kayaker -- nice!

Two comments: 

(1) From your description above, it sounds like that if you pass a `ComponentMask` with M out of N (and M<=N) components set, then the `Function` object would only have to have M components. Is this correct? If so, this is different from the design we have in most other places where the `Function` object still has to have N components, but the N-M components that are not set in the mask are simply ignored.

(2) You seem to overload several function signatures that now just forward to the new function with a `ComponentMask()` argument. If so, why not just provide the `ComponentMask` argument with a default value? This way you won't need all of those overloads.

**Comment by class4kayaker** [comment]
For 1), that is correct. I did look note the variation, but as I mentioned, I was thinking of the example case for Aspect. In that case, the functions to be interpolated could only reasonably return values for a particular set of components, and the expected number and location of those components could not reasonably be expected to be available to the interpolating function. One possible alternate approach would be to change to the standard ComponentMask approach and add a standard wrapper for functions that would move values from another function into the correct component locations if no such function currently exists. If such a function does exist, then it would make sense to change to make use of that, but I did not know of such a function. That approach might be more robust and would be useful for other cases so I will try looking into it.

For 2), that does make sense, and I will make that change.

I currently think that it would make the most sense to get this merged without handling the case of interpolating onto interpolating components of systems with non-interpolating FEs, and then handle that as a separate issue.

**Comment by bangerth** [comment]
Such function "adapters" already exist. See for example several of the *FunctionFrom*Object classes here: https://dealii.org/8.4.0/doxygen/deal.II/classFunction.html

Regarding the merge: yes, I'm all for incremental steps. They are so much easier to write, test, and review.

**Comment by class4kayaker** [comment]
I have made the desired modifications, and added another test. I believe that this is ready for review.

**Comment by class4kayaker** [comment]
The ComponentMask for VectorTools::interpolate is complete, and ready for review.

File: include/deal.II/numerics/vector_tools.h [hunk 1]
```diff
@@ -596,7 +597,8 @@ namespace VectorTools
   template <typename VectorType, typename DoFHandlerType>
   void interpolate (const DoFHandlerType                                   &dof,
                     const Function<DoFHandlerType::space_dimension,typename VectorType::value_type> &function,
-                    VectorType                                             &vec);
+                    VectorType                                             &vec,
+                    const ComponentMask                &component_mask = ComponentMask());
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.h]: please indent to the same level as the other arguments

File: include/deal.II/numerics/vector_tools.templates.h [hunk 2]
```diff
@@ -247,9 +255,12 @@ namespace VectorTools
                     {
                       const unsigned int component
                         = fe[fe_index].system_to_component_index(i).first;
-                      const unsigned int rep_dof=dof_to_rep_index_table[fe_index][i];
-                      vec(dofs_on_cell[i])
-                        = function_values_system[fe_index][rep_dof](component);
+                      if (component_mask[component] == true)
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: oh, neat, there really isn't very much to it, is there :-)

File: tests/dofs/interpolate_q_system_mask_02.output [hunk 3]
```diff
@@ -0,0 +1,93 @@
+
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 1, P_0, rel. error=0
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 2, P_0, rel. error=0
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 1, P_1, rel. error=0
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 2, P_1, rel. error=0
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 1, P_2, rel. error=0.000425
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 2, P_2, rel. error=0.000425
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 1, P_3, rel. error=0.00125
+DEAL::FESystem<1>[FE_Q<1>(1)^2-FE_Q<1>(2)], Mask 2, P_3, rel. error=0.00125
```
Comment (bangerth) [inline tests/dofs/interpolate_q_system_mask_02.output]: did you check that these errors are expected to be nonzero?

**Comment by bangerth** [comment]
Just one little comment and a question. Other than that, this looks good. Can you add an entry to `doc/news/changes.h` as well?

File: tests/dofs/interpolate_q_system_mask_02.output (hunk 3 above)
Comment (class4kayaker) [inline tests/dofs/interpolate_q_system_mask_02.output]: The test still includes the assertion from interpolate_q_system_01, and I checked these against the unmasked output (interpolate_q_system_01, interpolate_q_system_02) for where the nonzero values started.

Comment (bangerth) [inline tests/dofs/interpolate_q_system_mask_02.output]: OK, great. Thanks for checking.

**Comment by bangerth** [comment]
OK to merge if (i) you rebase to master and resolve the merge conflict, (ii) squash the different commits into one, or one each for code changes/changelog entry/tests.

**Pull request merged**